mlflow/tracing/export/mlflow.py

```diff
@@ -70,24 +76,159 @@ def export(self, root_spans: Sequence[ReadableSpan]):
-            self._client._upload_trace_data(trace.info, trace.data)
-            self._client._upload_ended_trace_info(trace.info)
-        except Exception as e:
-            # avoid silent failures
             _logger.warning(
-                f"Failed to log trace to MLflow backend: {e}",
+                f"{self._error_msg} Error: {e}. For full traceback, set logging level to debug.",
                 exc_info=_logger.isEnabledFor(logging.DEBUG),
             )
+
+
+class AsyncTraceExportQueue:
+    """
+    This is a queue based run data processor that queue incoming data and process it using a single
+    worker thread. This class is used to process traces saving in async fashion.
+    """
+
+    def __init__(self, client) -> None:
+        self._queue: Queue[Task] = Queue()
+        self._client = client
+        self._lock = threading.RLock()
+
+        # A thread event that indicates the logging queue stop processing task.
+        self._stop_event = threading.Event()
+        self._is_activated = False
+        self._unprocessed_tasks = set()
+
+    def put(self, task: Task) -> None:
+        """Put a new task to the queue for processing."""
+        if not self.is_active():
+            self.activate()
+
+        # If stop even is set, we should wait for the queue to be drained before putting the task.
+        if self._stop_event.is_set():
+            self._stop_event.wait()
+
+        self._unprocessed_tasks.add(task)
+        self._queue.put(task)
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously process incoming tasks in the queue until the stop event is set.
+        When the stop event is set, the loop will drain the remaining tasks in the queue.
+        """
+        while not self._stop_event.is_set():
+            self._handle_task()
+        while not self._queue.empty() or self._unprocessed_tasks:
+            self._handle_task()
+
+    def _handle_task(self) -> None:
+        """Process the given task in the running runs queues."""
+        try:
+            task = self._queue.get(timeout=1)
+        except Empty:
+            return
+
+        def _handle(task):
+            task.handle()
+            self._unprocessed_tasks.remove(task)
+
+        self._trace_logging_worker_threadpool.submit(_handle, task)
+
+    def activate(self) -> None:
+        """Activates the async logging queue"""
+        with self._lock:
+            if self._is_activated:
+                return
+
+            self._set_up_logging_thread()
+            # Registering an atexit callback to ensure that any remaining log data
+            # is flushed before the program exits.
+            atexit.register(self._at_exit_callback)
+            self._is_activated = True
+
+    def is_active(self) -> bool:
+        return self._is_activated
+
+    def _set_up_logging_thread(self) -> None:
+        """Sets up the logging thread."""
+        with self._lock:
+            self._trace_logging_thread = threading.Thread(
```

Comment:
Since activate can be called after a flush call, can we add a quick safety check here to ensure that we only start the logging thread up if it's undefined or is not active?

`if self._trace_logging_thread is None or not self._trace_logging_thread.is_alive():`

Reply:
I think it doesn't happen because we have this condition within `activate`
```
with self._lock:
    if self._is_activated:
        return
```
Only one thread should activate the threads and others will return.